pinot-core/src/main/java/org/apache/pinot/core/data/table/IndexedTable.java

```diff
@@ -0,0 +1,163 @@
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.core.data.table;
+
+import com.google.common.base.Preconditions;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.List;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import javax.annotation.Nonnull;
+import org.apache.commons.collections.CollectionUtils;
+import org.apache.pinot.common.request.AggregationInfo;
+import org.apache.pinot.common.request.SelectionSort;
+import org.apache.pinot.common.utils.DataSchema;
+import org.apache.pinot.core.data.order.OrderByUtils;
+import org.apache.pinot.core.query.aggregation.function.AggregationFunction;
+import org.apache.pinot.core.query.aggregation.function.AggregationFunctionUtils;
+
+
+/**
+ * {@link Table} implementation for aggregating TableRecords based on combination of keys
+ */
+public class IndexedTable implements Table {
+
+  // When table reaches max capacity, we will allow 20% more records to get inserted (bufferedCapacity)
+  // If records beyond bufferedCapacity are received, the table will undergo sort and evict upto evictCapacity (10% more than capacity)
+  // This is to ensure that for a small number beyond capacity, a fair chance is given to all records which have the potential to climb up the order
+  /** Factor used to add buffer to maxCapacity of the Collection used **/
+  private static final double BUFFER_FACTOR = 1.2;
+  /** Factor used to decide eviction threshold **/
+  private static final double EVICTION_FACTOR = 1.1;
+
+  private List<Record> _records;
+  private ConcurrentMap<Record, Integer> _lookupTable;
+  private ReentrantReadWriteLock _readWriteLock;
+
+  private DataSchema _dataSchema;
+  private List<AggregationInfo> _aggregationInfos;
+  private List<AggregationFunction> _aggregationFunctions;
+  private List<SelectionSort> _orderBy;
+  private int _evictCapacity;
+  private int _bufferedCapacity;
+
+  @Override
+  public void init(@Nonnull DataSchema dataSchema, List<AggregationInfo> aggregationInfos, List<SelectionSort> orderBy,
+      int maxCapacity) {
+    _dataSchema = dataSchema;
+    _aggregationInfos = aggregationInfos;
+    _orderBy = orderBy;
+
+    _bufferedCapacity = (int) (maxCapacity * BUFFER_FACTOR);
+    _evictCapacity = (int) (maxCapacity * EVICTION_FACTOR);
+
+    _records = new ArrayList<>(_bufferedCapacity);
```

Comment:
Does this not need to be concurrent?

Reply:
Hmm, yes. I'm not sure there is a concurrent list implementation available. I made it a synchronized list for thread safety, but now the add() to list is synchronized.. Any suggestions?